File: BOJ/2644.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;
int n, m, x, y, p1, p2, dis[102] = {0};
vector<int> v[102];

void input(){
    cin >> n >> p1 >> p2 >> m;
    for(int i=0; i<m; i++){
        cin >> x >> y;
        v[x].push_back(y);
        v[y].push_back(x);
    }
}

void bfs(int s, int e){
    queue<int> q;
    q.push(s);
    while(!q.empty()){
        int x = q.front(); q.pop();
        if(x==e){
            cout << dis[x] << '\n';
            return;
        }
        
        for(int i=0; i<v[x].size(); i++){ //인접한 노드로 이동
            int y = v[x][i];
            if(dis[y] == 0){ //방문하지 않은 곳
                q.push(y);
                dis[y] = dis[x] + 1;
            }
        }
    }
    
    cout << "-1\n"; //이 안에서 안나오면 가족이 아님
}

int main(){
    ios::sync_with_stdio(0); cin.tie(0); cout.tie(0);
    input();
    bfs(p1, p2);
    return 0;
}

File: BOJ/2644.swift
import Foundation

let n = Int(readLine()!)!
let people = readLine()!.split(separator: " ").map{ Int(String($0))! }
let one = people[0]
let two = people[1]
let m = Int(readLine()!)!
var graph = [[Int]](repeating: [], count: n+1)
var visited = [Int](repeating: -1, count: n+1)
for _ in 0..<m {
    let input = readLine()!.split(separator: " ").map{ Int(String($0))! }
    graph[input[0]].append(input[1])
    graph[input[1]].append(input[0])
}
func BFS() -> Int {
    var queue = [one]
    var front = 0
    visited[one] = 0
    
    while queue.count > front {
        let x = queue[front]
        front += 1
        
        if x == two {
            return visited[x]
        }
        
        for next in graph[x] {
            guard visited[next] == -1 else { continue }
            visited[next] = visited[x] + 1
            queue.append(next)
        }
    }
    return -1
}

print(BFS())
